src/V4Router.sol

```diff
@@ -0,0 +1,168 @@
+import {BalanceDelta} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {Currency, CurrencyLibrary} from "@uniswap/v4-core/src/types/Currency.sol";
+import {TickMath} from "@uniswap/v4-core/src/libraries/TickMath.sol";
+import {TransientStateLibrary} from "@uniswap/v4-core/src/libraries/TransientStateLibrary.sol";
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+
+import {PathKey, PathKeyLib} from "./libraries/PathKey.sol";
+import {CalldataDecoder} from "./libraries/CalldataDecoder.sol";
+import {IV4Router} from "./interfaces/IV4Router.sol";
+import {BaseActionsRouter} from "./base/BaseActionsRouter.sol";
+import {DeltaResolver} from "./base/DeltaResolver.sol";
+import {Actions} from "./libraries/Actions.sol";
+
+/// @title UniswapV4Router
+/// @notice Abstract contract that contains all internal logic needed for routing through Uniswap V4 pools
+/// @dev the entry point to executing actions in this contract is calling `BaseActionsRouter._executeActions`
+/// An inheriting contract should call _executeActions at the point that they wish actions to be executed
+abstract contract V4Router is IV4Router, BaseActionsRouter, DeltaResolver {
+    using PathKeyLib for PathKey;
+    using CalldataDecoder for bytes;
+    using TransientStateLibrary for IPoolManager;
+
+    constructor(IPoolManager poolManager) BaseActionsRouter(poolManager) {}
+
+    // TODO native support !!
+    function _handleAction(uint256 action, bytes calldata params) internal override {
+        // swap actions and payment actions in different blocks for gas efficiency
+        if (action < Actions.SETTLE) {
+            if (action == Actions.SWAP_EXACT_IN) {
+                _swapExactInput(abi.decode(params, (IV4Router.ExactInputParams)));
+            } else if (action == Actions.SWAP_EXACT_IN_SINGLE) {
+                _swapExactInputSingle(abi.decode(params, (IV4Router.ExactInputSingleParams)));
+            } else if (action == Actions.SWAP_EXACT_OUT) {
+                _swapExactOutput(abi.decode(params, (IV4Router.ExactOutputParams)));
+            } else if (action == Actions.SWAP_EXACT_OUT_SINGLE) {
+                _swapExactOutputSingle(abi.decode(params, (IV4Router.ExactOutputSingleParams)));
+            } else {
+                revert UnsupportedAction(action);
+            }
+        } else {
+            if (action == Actions.SETTLE_ALL) {
+                // equivalent: abi.decode(params, (Currency))
+                Currency currency;
+                assembly ("memory-safe") {
+                    currency := calldataload(params.offset)
+                }
+
+                int256 delta = poolManager.currencyDelta(address(this), currency);
+                if (delta > 0) revert InvalidDeltaForAction();
+
+                // TODO support address(this) paying too
+                // TODO should it have a maxAmountOut added slippage protection?
```

Comment:
my thought is that slippage protection should be with the swap commands? why do we also need slippage protection with settle commands?

i think we can make settle just take an amount, and if you're ok with the full settle amount then you pass in a constant/sentinel

Reply:
Its to support split route trades. Lets say youre trading 1000 DAI for USDC, where you want at least 998 USDC
- 500 DAI -> USDT -> USDC
- 500 DAI -> USDC

Then you would do a command thats like "take all USDC, and check we're receiving at least 998".

Reply:
We implement the same style of logic in UR using the sweep command e.g. [here](https://github.com/Uniswap/universal-router/blob/4cf96077ba21dbbbbbc8ffa8c549648f69623fbc/contracts/modules/Payments.sol#L80)

Reply:
And we do it with the unwrapETH command to do the same check for ETH output trades e.g. [here](https://github.com/Uniswap/universal-router/blob/4cf96077ba21dbbbbbc8ffa8c549648f69623fbc/contracts/modules/Payments.sol#L130)

Reply:
but surely you're also checking slippage from USDT to USDC and from DAI to USDC? or are you ignoring those checks until the end? My concern is that we're doing lots of slippage checks and the ones in the middle then dont even matter?

Reply:
> but surely you're also checking slippage from USDT to USDC and from DAI to USDC? or are you ignoring those checks until the end? My concern is that we're doing lots of slippage checks and the ones in the middle then dont even matter?

yeah the ones in the middle dont matter. the user doesnt care if one half of the trade gets hit with more slippage than the other - theres just an output amount that they want from the trade in total and thats it

Reply:
>My other thought process then also is just if posm & routers take/settle commands looks different now

yeah i could do it as a different cmmand or something depending how the other commands end up